src/target/linux/software_breakpoint.rs

```diff
@@ -0,0 +1,95 @@
+//! Bundles functionalities related to software breakpoints
+//! Software Breakpoints work by overwritting the target program's memory;
+//! replacing an instruction with one that causes a signal to be raised by the
+//! cpu.
+
+use nix::sys::ptrace;
+use nix::unistd::Pid;
+use std::cell::Cell;
+use std::rc::Rc;
+const INT3: libc::c_long = 0xcc;
+
+#[derive(Debug, Clone)]
+pub struct Breakpoint {
+    /// The address at which the debugger should insert this breakpoint
+    pub addr: usize,
+    /// The original instruction overwritten by the breakpoint
+    pub(super) shadow: i64,
+    pid: Pid,
+    user_enabled: Rc<Cell<bool>>,
+}
+
+impl Breakpoint {
+    /// Set a breakpoint at a given address
+    pub(crate) fn new(addr: usize, pid: Pid) -> Result<Self, BreakpointError> {
+        let shadow = ptrace::read(pid, addr as *mut _)?;
```

Comment:
If we use `target::ReadMemory` here, I think we can make this module generic for `x86_64` and OS-independent.